liquidator/liquidator.js

```diff
@@ -1,37 +1,50 @@
-const argv = require("minimist")(process.argv.slice(), { string: ["address"] });
-
+// When running this script it assumes to have enough tokens and allowance from the unlocked truffle wallet.
+// Future versions will deal with generating additional synthetic tokens from EMPs as the bot needs.
 const { ExpiringMultiPartyClient } = require("../financial-templates-lib/ExpiringMultiPartyClient.js");
-const { delay } = require("../financial-templates-lib/delay.js");
 
 const ExpiringMultiParty = artifacts.require("ExpiringMultiParty");
 
-// TODO: Figure out a good way to run this script, maybe with a wrapper shell script.
-// Currently, you can run it with `truffle exec ../liquidator/liquidator.js --address=<address>` *from the core
-// directory*.
-async function run() {
-  const client = new ExpiringMultiPartyClient(ExpiringMultiParty.abi, web3, argv.address);
-  client.start();
-  while (true) {
-    try {
-      console.log("Polling");
-      // Steps:
-      // Get most recent price from a price feed.
-      // Call client.getUnderCollateralizedPositions()
-      // Acquire synthetic tokens somehow. v0: assume the bot holds on to them.
-      // Liquidate any undercollateralized positions!
-      // Withdraw money from any liquidations that are expired or DisputeFailed.
-    } catch (error) {
-      console.log("Poll error:", error);
-    }
-    await delay(Number(10_000));
+class Liquidator {
+  constructor(abi, web3, empAddress) {
+    this.empAddress = empAddress;
+    web3.eth.getAccounts().then(accounts => {
+      this.account = accounts[0];
+      console.log(this.account);
+    });
+
+    // Expiring multiparty contract to read contract state
+    this.empClient = new ExpiringMultiPartyClient(abi, web3, empAddress);
+    this.empClient.start();
+
+    // Instance of the expiring multiparty to perform on-chain liquidations
+    this.empContract = new web3.eth.Contract(abi, empAddress);
   }
+
+  // Queries underCollateralized positions and performs liquidations against any under collateralized positions.
+  queryAndLiquidate = async priceFeed => {
+    console.log("Checking for under collateralized positions at the price", priceFeed);
+    await this.empClient._update();
+    const underCollateralizedPositions = this.empClient.getUnderCollateralizedPositions(priceFeed);
+    console.log("UNDER COLLATERALIZED POSITIONS:", underCollateralizedPositions);
+
+    for (let i = 0; i < underCollateralizedPositions.length; i++) {
+      console.log("Liquidating sponsor", underCollateralizedPositions[i].sponsor);
+      // Create the liquidation transaction
+      // TODO calculate the amountToLiquidate as a function of the total collateral within the position
+      // and the current price of the collateral. This will require knowing how much the collateral and the
+      // synthetic are worth.
+      this.empContract.methods
```

Comment:
So if I understand correctly, this loop will fire off all the liquidations without waiting for the previous one to finish? I think that's really good!

Just wondering if we should wait for all of them to finish before returning from this async `queryAndLiquidate` function though?